### PR TITLE
Move the definitions of archimedean structures from ssrnum.v to archimedean.v

### LINKED ISSUE #1236: renaming: `archiDomain`, etc.
https://github.com/math-comp/math-comp/blob/36dbdd6bb3c5e2a46c075ab0344920466eaf4fea/mathcomp/algebra/ssrnum.v#L427

`archiDomain` -> `archiRealDomain`
similarly:
`archiField` -> `archiRealField`

see conversation item https://github.com/math-comp/math-comp/pull/1235#issuecomment-2199976942

### PR DESCRIPTION
**Pull request by pi8027**
##### Motivation for this change

In #682, we added some new archimedean structures, deprecated them in `ssrnum.v`, and asked the users to require a new file called `archimedean.v` to use them. This PR moves the definitions of archimedean structures and the declarations depending on them from `ssrnum.v` to `archimedean.v` to clean up the deprecation notations.

This PR also renames `archiDomainType` and `archiFieldType` to `archiRealDomainType` and `archiRealFieldType`, respectively, hence closes #1236.

##### PR overlays

- math-comp/Abel#93
- math-comp/analysis#1244
- coq-community/apery#23
- math-comp/real-closed#61

##### Minimal TODO list

- [x] added corresponding entries in `CHANGELOG_UNRELEASED.md`

- [x] added corresponding documentation in the headers
- [x] tried to abide by the [contribution guide](https://github.com/math-comp/math-comp/blob/master/CONTRIBUTING.md)
- [x] this PR contains an optimum number of meaningful commits

See [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-creating-and-review-PRs) for details.

##### Compatibility with MathComp 1.X

- ~[ ] I added the label `TODO: MC-1 port` to make sure someone ports this PR to
      the `mathcomp-1` branch **or** I already opened an issue or PR (please cross reference).~

##### Automatic note to reviewers

Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-creating-and-review-PRs).

### CODE REVIEW

**Comment by pi8027** [comment]
I will add a commit to address #1236 later.

File: mathcomp/algebra/archimedean.v [hunk 1]
```diff
@@ -94,20 +153,45 @@ without loss x_ge0: x Rx / x >= 0.
...
+
+Lemma is_intE n : true = (0 <= n) || (0 <= - n).
+Proof. by case: n. Qed.
+
+End intArchimedean.
+End intArchimedean.
+
+#[export]
+HB.instance Definition _ := Num.NumDomain_isArchimedean.Build int
+  intArchimedean.truncP intArchimedean.is_natE intArchimedean.is_intE.
```
Comment (pi8027) [inline mathcomp/algebra/archimedean.v:194]: I'm not sure if this is the right place to declare this instance, but `Znat_def` and `ZnatP` in the `Theory` module below depend on it.

File: mathcomp/algebra/archimedean.v [hunk 2]
```diff
@@ -53,34 +53,93 @@ Local Open Scope ring_scope.
...
+  nat_num_subdef : pred R;
+  int_num_subdef : pred R;
+  trunc_subproof :
+    forall x,
+      if 0 <= x then (trunc_subdef x)%:R <= x < (trunc_subdef x).+1%:R
+      else trunc_subdef x == 0%N;
+  nat_num_subproof : forall x, nat_num_subdef x = ((trunc_subdef x)%:R == x);
+  int_num_subproof :
+    forall x, int_num_subdef x = nat_num_subdef x || nat_num_subdef (- x);
+}.
```
Comment (pi8027) [inline mathcomp/algebra/archimedean.v:69]: I'm unsure if we should keep the `_subdef` prefix here.

File: mathcomp/algebra/archimedean.v [hunk 3]
```diff
@@ -94,20 +153,45 @@ without loss x_ge0: x Rx / x >= 0.
 by exists (Posz (trunc x)); rewrite addrC -intS -!pmulrn trunc_itv.
 Qed.
 
-End ArchiNumDomain.
-
-End Internals.
+Definition floor x := sval (floor_subproof x).
+Definition ceil x := - floor (- x).
+Definition archi_bound x := (trunc `|x|).+1.
```
Comment (pi8027) [inline mathcomp/algebra/archimedean.v:158]: Since `archimedean.v` depends on `ssrint`, now we can consider including (at least some of) these operators in the mixin instead of constructing them from `trunc`.

**Comment by pi8027** [comment]
@affeldt-aist Can you take a look at analysis?

**Comment by affeldt-aist** [comment]
> @affeldt-aist Can you take a look at analysis?

you mean check if the changes are compatible with mathcomp-analysis? will do asap

**Comment by pi8027** [comment]
@affeldt-aist I'm rather asking for a fix because we already know that analysis is broken (see CI).

File: mathcomp/algebra/all_algebra.v [hunk 4]
```diff
@@ -7,6 +7,7 @@ Require Export polydiv.
 Require Export polyXY.
 Require Export qpoly.
 Require Export ssrint.
+Require Export archimedean.
```
Comment (pi8027) [inline mathcomp/algebra/all_algebra.v:10]: `Require Export`s in this file are not in the dependency order.

**Comment by affeldt-aist** [comment]
> @affeldt-aist I added your PR as an overlay but analysis is still broken.

isn't that the result of a renaming?
```
> File "./Rstruct.v", line 295, characters 28-61:
> Error: The reference Num.RealField_isArchimedean.Build was not found
> in the current environment.
```

**Comment by pi8027** [comment]
@affeldt-aist Yes, but that was deprecated in MathComp 2.1.0.

**Comment by affeldt-aist** [comment]
Yes, indeed, I missed this one in the flow of warnings, I'll take care of it now

**Comment by pi8027** [comment]
I still need to remove `Qint`, `Cint`, etc., and change at least Abel and Apery accordingly.

File: mathcomp/algebra/archimedean.v (hunk 3 above)
Comment (proux01) [reply mathcomp/algebra/archimedean.v:158]: Sounds a good idea but a larger change, maybe this should be done in a further PR.

Comment (pi8027) [reply mathcomp/algebra/archimedean.v:158]: Agreed.

**Comment by pi8027** [comment]
Done modulo documentation

File: mathcomp/algebra/archimedean.v [hunk 5]
```diff
@@ -595,23 +688,88 @@ End ArchiClosedFieldTheory.
...
+Module ArchiField.
+#[deprecated(since="mathcomp 2.3.0",
+  note="Use Num.ArchiRealField.type instead.")]
+Notation type := ArchiRealField.type.
+#[deprecated(since="mathcomp 2.3.0",
+  note="Use Num.ArchiRealField.copy instead.")]
+Notation copy T C := (ArchiRealField.copy T C).
+#[deprecated(since="mathcomp 2.3.0", note="Use Num.ArchiRealField.on instead.")]
+Notation on T := (ArchiRealField.on T).
+End ArchiField.
```
Comment (pi8027) [inline mathcomp/algebra/archimedean.v:766]: What is the easiest way to deprecate an HB class?

**Comment by pi8027** [comment]
@Tragicus @CohenCyril FYI, this PR may break some of your backport PRs from Abel. See also math-comp/Abel#93.

**Comment by pi8027** [comment]
This PR and all the overlay PRs listed above are ready.

**Comment by Tragicus** [comment]
If you are ready, I think you can ignore us and I will rebase when you merge.

**Comment by pi8027** [comment]
I will rebase this PR after merging #1245.

**Comment by pi8027** [comment]
@CohenCyril This PR and its only overlay PR left (math-comp/Abel#93) are both ready. It's also blocking a potential improvement (simplification) of the archimedean mixin. Do you have time to take a look soon?

**Comment by pi8027** [comment]
I removed the overlay.

File: mathcomp/algebra/archimedean.v (hunk 2 above)
Comment (CohenCyril) [reply mathcomp/algebra/archimedean.v:65]: I think it is not necessary anymore

Comment (CohenCyril) [reply mathcomp/algebra/archimedean.v:65]: and same for `trunc_subproof`

Comment (pi8027) [reply mathcomp/algebra/archimedean.v:65]: Since I will likely change the mixin anyway, I prefer to address this in a separate PR.

**Review by CohenCyril — APPROVED** [review]
LGTM
NB: We should definitely open an issue on HB about deprecating HB structures/factories. @pi8027 since you needed it could you summarize the desired features in such an issue?

**Comment by pi8027** [comment]
> We should definitely open an issue on HB about deprecating HB structures/factories. @pi8027 since you needed it could you summarize the desired features in such an issue?

Done: math-comp/hierarchy-builder#436

**Comment by pi8027** [comment]
Since Cyril approved this PR, let me merge it and proceed to the next PR. Thanks @proux01 @affeldt-aist @CohenCyril for the help!
